src/main/java/org/embulk/input/marketo/delegate/MarketoBaseBulkExtractInputPlugin.java

```diff
@@ -0,0 +1,199 @@
+    {
+        ConfigDiff configDiff = super.buildConfigDiff(task, schema, taskCount, taskReports);
+        Long currentLatestFetchTime = 0L;
+        Set<String> latestUIds = null;
+        for (TaskReport taskReport : taskReports) {
+            Long latestFetchTime = taskReport.get(Long.class, LATEST_FETCH_TIME);
+            if (latestFetchTime == null) {
+                continue;
+            }
+            if (currentLatestFetchTime < latestFetchTime) {
+                currentLatestFetchTime = latestFetchTime;
+                latestUIds = taskReport.get(Set.class, LATEST_UID_LIST);
+            }
+        }
+        if (!currentLatestFetchTime.equals(0L)) {
+            configDiff.set(LATEST_FETCH_TIME, currentLatestFetchTime);
+            DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
+            configDiff.set("from_date", df.format(new Date(currentLatestFetchTime)));
+            configDiff.set(LATEST_UID_LIST, latestUIds);
+        }
+        return configDiff;
+    }
+
+    @Override
+    public TaskReport ingestServiceData(T task, RecordImporter recordImporter, int taskIndex, PageBuilder pageBuilder)
+    {
+        InputStream extractedStream = getExtractedStream(task, pageBuilder.getSchema());
+        return importRecordFromFile(task, extractedStream, recordImporter, pageBuilder);
+    }
+
+    protected TaskReport importRecordFromFile(T task, InputStream inputStream, RecordImporter recordImporter, PageBuilder pageBuilder)
+    {
+        Set<String> latestUids = task.previousUIds();
+        TaskReport taskReport = Exec.newTaskReport();
+        int imported = 0;
+        Timestamp currentTimestamp = null;
+        if (task.getLatestFetchTime().isPresent()) {
+            currentTimestamp = Timestamp.ofEpochMilli(task.getLatestFetchTime().get());
+        }
+        TimestampParser timestampParser = new TimestampParser(MarketoUtils.ISO_8601_FORMAT, DateTimeZone.UTC);
+        try (LineDecoder lineDecoder = new LineDecoder(new InputStreamFileInput(task.getBufferAllocator(), inputStream), task)) {
+            CsvTokenizer csvTokenizer = new CsvTokenizer(lineDecoder, task);
+            if (!csvTokenizer.nextFile()) {
+                throw new DataException("Can't read extract input stream");
+            }
+            csvTokenizer.nextRecord();
+            List<String> headers = new ArrayList<>();
+            while (csvTokenizer.hasNextColumn()) {
+                headers.add(csvTokenizer.nextColumn());
+            }
+            while (csvTokenizer.nextRecord()) {
+                List<String> values = new ArrayList<>();
+                while (csvTokenizer.hasNextColumn()) {
+                    values.add(csvTokenizer.nextColumnOrNull());
+                }
+                if (imported >= PREVIEW_RECORD_LIMIT && Exec.isPreview()) {
+                    break;
+                }
+                final Map<String, String> kvMap = MarketoUtils.zip(headers, values);
```

Comment:
I expect the CVS file will be large and this seems to be inefficient when the number of records is large enough. I think you can just build up the map<k,v> directly from reading the columns of each line using indices. This saves another loop over the headers & values to build up the map.

Reply:
Yes, I thought the same too, but the returned fields and schema column may not be in the same order. Since there is no guaranteed that Marketo will return headers name in the same order with the field we sent to them. So I think it better if we keep it this way